cumulus/client/bootnodes/src/discovery.rs

```diff
@@ -0,0 +1,460 @@
+		if self.key_being_discovered.is_none() &&
+			self.pending_responses.is_empty() &&
+			self.find_node_queries.is_empty()
+		{
+			// No need to start discovey again if the previous attempt succeeded.
+			if self.succeeded {
+				info!(
+					target: LOG_TARGET,
+					"Parachain bootnode discovery on the relay chain DHT succeeded",
+				);
+				Ok(false)
+			} else if self.attempts_left > 0 {
+				// No discovery in progress and we have attempts left, start discovery.
+				self.attempts_left -= 1;
+				self.start_discovery().await?;
+				Ok(true)
+			} else {
+				warn!(
+					target: LOG_TARGET,
+					"Failed to discover parachain bootnodes on the relay chain DHT after {} attempts, \
+					 giving up",
+					MAX_DISCOVERY_ATTEMPTS,
+				);
+				Ok(false)
+			}
+		} else {
+			// Discovery is already in progress, just continue the event loop.
+			Ok(true)
+		}
+	}
+
+	fn request_bootnode(&mut self, peer_id: PeerId) {
+		trace!(
+			target: LOG_TARGET,
+			"Requesting parachain bootnode from the relay chain {peer_id:?}",
+		);
+
+		let (tx, rx) = oneshot::channel();
+
+		self.relay_chain_network.start_request(
+			peer_id,
+			self.paranode_protocol_name.clone(),
+			self.para_id_scale_compact.clone(),
+			None,
+			tx,
+			IfDisconnected::TryConnect,
+		);
+
+		self.pending_responses.push(async move { (peer_id, rx.await) }.boxed());
+	}
+
+	fn handle_providers(&mut self, providers: Vec<PeerId>) {
+		debug!(
+			target: LOG_TARGET,
+			"Found parachain bootnode providers on the relay chain: {providers:?}",
+		);
+
+		for peer_id in providers {
+			if peer_id == self.relay_chain_network.local_peer_id() {
+				continue;
+			}
+
+			// libp2p may yield the same provider multiple times; skip if we alredy queried it.
+			if self.direct_requests.contains(&peer_id) || self.find_node_queries.contains(&peer_id)
+			{
+				continue;
+			}
+
+			// Directly request a bootnode from the peer without performing a `FIND_NODE` query
+			// first. With litep2p backend this will likely succeed, because cached provider
+			// addresses are automatically added to the transport manager known addresses list.
+			//
+			// With libp2p backend, or if the remote did not return the cached addresses of the
+			// provider, the request will fail and we will perform a `FIND_NODE` query.
+			self.direct_requests.insert(peer_id);
+			self.request_bootnode(peer_id);
+		}
+	}
+
+	fn handle_response(
+		&mut self,
+		peer_id: PeerId,
+		res: Result<Result<(Vec<u8>, ProtocolName), RequestFailure>, oneshot::Canceled>,
+	) {
+		let direct_request = self.direct_requests.remove(&peer_id);
+
+		let response = match res {
+			Ok(Ok((payload, _))) => match Response::decode(payload.as_slice()) {
+				Ok(response) => response,
+				Err(e) => {
+					warn!(
+						target: LOG_TARGET,
+						"Failed to decode parachain bootnode response from {peer_id:?}: {e}",
+					);
+					return;
+				},
+			},
+			Ok(Err(e)) => {
+				if direct_request {
```

Comment:
There are a few cases where we could skip this process (ie UnsupportedProtocols and UnknownProtocol), and maybe a case where we should retry the request (Timeout)?

Reply:
We could also add a test for both backends, where we make a call on this protocol and catch the expected result on empty addresses. Then we assert the error should remain constant over time.
If something changes in the backend, we'll get notified by this test and we'll either adjust the backend or the impl

Reply:
Yes, yes, but libp2p & litep2p report the needed error (missing address in the cache that should only interest us) differently, and with libp2p there is no way to distinguish between network error and a missing address. So not to go into the peculiarities of internal implementations and also make it future-proof if the error reporting changes in the backends, I just always trigger the lookup. This shouldn't incur a significant overhead, as it is only done once (when the node is started.)

Let me know if this works for you.